### PR TITLE
IGNITE-16705 Resolve Compute futures that had no chance to be executed on node stop

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-16705

### CODE REVIEW

**Review by SammyVimes — APPROVED** [review]
LGTM

File: modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Maintains a collection of in-flight {@link CompletableFuture}s (i.e. futures that are not yet completed) to later
+ * have a possibility to process them somehow (for example, cancel them all if we know for sure that they will
+ * never be completed).
+ */
+public class InFlightFutures implements Iterable<CompletableFuture<?>> {
+    private final Set<CompletableFuture<?>> inFlightFutures = Collections.newSetFromMap(new ConcurrentHashMap<>());
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java:32]: ConcurrentHashMap.newKeySet() WDYT?

File: modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    /**
+     * Registers a future in the in-flight futures collection. When it completes (either normally or exceptionally),
+     * it will be removed from the collection.
+     *
+     * @param future the future to register
+     */
+    public void registerFuture(CompletableFuture<?> future) {
+        future.whenComplete((result, ex) -> inFlightFutures.remove(future));
```
Comment (sanpwc) [inline modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java:41]: And what about pipelining? The ordering of independent completion stages isn't guaranteed, so it's possible to remove the future while one of it's stages is still in progress. E.g.
```
CompletableFuture<R> future = startLocalExecution(jobClass, args);
inFlightFutures.registerFuture(future);
...
future.thenAcceptAsync()
```

**Review by sanpwc — COMMENTED** [review]
All in all, the patch looks fine to me. It's however an optimization, so it's still required to both protect critical sections inside futures with busy locks and handle NodeStoppingException.

File: modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java (hunk 1 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java]: Thanks, it's better this way, switched to `ConcurrentHashMap#newKeySet()`

File: modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java (hunk 2 above)
Comment (rpuch) [reply modules/core/src/main/java/org/apache/ignite/internal/future/InFlightFutures.java:40]: This PR is about making sure that a future does not get stuck due to our thread pool being stopped. It looks like the consequtive stages should be responsibility of those who add them.

There is also another problem: consequtive stages might be computationally heavy, and they can potentially block our 'job execution thread pool'. This needs further investigation, I've created an issue: https://issues.apache.org/jira/browse/IGNITE-16747

**Pull request merged**